Yafc.Model/Serialization/ValueSerializers.cs

```diff
@@ -334,6 +413,113 @@ public override void WriteToJson(Utf8JsonWriter writer, T? value) {
+        if (reader.TokenType is JsonTokenType.String or JsonTokenType.PropertyName) {
+            string value = reader.GetString()!;
+            if (value.StartsWith('!')) {
+                // Read the new dictionary-key format, typically "!Item.coal!normal"
+                string separator = Separator().Match(value).Value;
+                string[] parts = value.Split(separator);
+                if (parts.Length == 3) { // The parts are <empty>, target.typeDotName, and quality.name
+                    _ = Database.objectsByTypeName.TryGetValue(parts[1], out var obj);
+                    _ = Database.objectsByTypeName.TryGetValue("Quality." + parts[2], out var qual);
+                    if (obj is not T) {
+                        context.Error($"Factorio object '{parts[1]}' no longer exists. Check mods configuration.", ErrorSeverity.MinorDataLoss);
+                    }
+                    else if (qual is not Quality quality) {
+                        context.Error($"Factorio quality '{parts[2]}' no longer exists. Check mods configuration.", ErrorSeverity.MinorDataLoss);
+                    }
+                    else {
+                        return ObjectWithQuality.Get(obj as T, quality);
+                    }
+                }
+                // If anything went wrong reading the dictionary-key format, try the non-quality format instead.
+            }
+
+            // Read the old non-quality "Item.coal" format
+            return ObjectWithQuality.Get(objectSerializer.ReadFromJson(ref reader, context, owner), Quality.Normal);
+        }
+
+        if (reader.TokenType == JsonTokenType.StartObject) {
+            // Read the object-based quality format: { "target": "Item.coal", "quality": "Quality.normal" }
+            reader.Read(); // {
+            reader.Read(); // "target":
+            T? obj = objectSerializer.ReadFromJson(ref reader, context, owner);
+            reader.Read(); // target
+            reader.Read(); // "quality":
+            Quality? quality = qualitySerializer.ReadFromJson(ref reader, context, owner);
+            reader.Read(); // quality
+            // If anything went wrong, the error has already been reported. We can just return null and continue.
+            return quality == null ? null : ObjectWithQuality.Get(obj, quality);
+        }
+
+        if (reader.TokenType != JsonTokenType.Null) {
+            context.Error($"Unexpected token {reader.TokenType} reading a quality object.", ErrorSeverity.MinorDataLoss);
+        }
+
+        return null; // Input is JsonTokenType.Null, or couldn't parse the input
+    }
+
+    public override void WriteToJson(Utf8JsonWriter writer, IObjectWithQuality<T>? value) {
+        if (value == null) {
+            writer.WriteNullValue();
+            return;
+        }
+
+        // TODO: This writes values that can be read by older versions of Yafc. For consistency with quality objects written as dictionary keys,
```

Comment:
Is the backwards compatibility required? If people only upgrade (as we tend to fix issues fast), we could skip

Reply:
It was definitely required for me when I was switching between this branch and branches that couldn't read the `"entity": "!Entity.assembly-machine!normal"` format.

I think it'll be nice to let people switch back from 2.12 to 2.11, (in the same way we let people switch from 0.7.3 to 0.7.2) but by the time 2.14 or 2.15 is out, it'll be unnecessary.

Reply:
> switching between this branch and branches that couldn't read

True, let's keep it for a few versions